### ISSUE TITLE
Implicit coercions for ROW type not handled correctly

### ISSUE THREAD

**Issue by martint** [prestodb/presto#9014]
``` select row('a') = row('aa')```

fails with:

```
Query 20170920_184840_03951_gui64 failed: line 1:17: '=' cannot be applied to row(field0 varchar(1)), row(field0 varchar(2))
com.facebook.presto.sql.analyzer.SemanticException: line 1:17: '=' cannot be applied to row(field0 varchar(1)), row(field0 varchar(2))
	at com.facebook.presto.sql.analyzer.ExpressionAnalyzer$Visitor.getOperator(ExpressionAnalyzer.java:1147)
	at com.facebook.presto.sql.analyzer.ExpressionAnalyzer$Visitor.visitComparisonExpression(ExpressionAnalyzer.java:446)
	at com.facebook.presto.sql.analyzer.ExpressionAnalyzer$Visitor.visitComparisonExpression(ExpressionAnalyzer.java:282)
	at com.facebook.presto.sql.tree.ComparisonExpression.accept(ComparisonExpression.java:71)
	at com.facebook.presto.sql.tree.StackableAstVisitor.process(StackableAstVisitor.java:26)
	at com.facebook.presto.sql.analyzer.ExpressionAnalyzer$Visitor.process(ExpressionAnalyzer.java:301)
	at com.facebook.presto.sql.analyzer.ExpressionAnalyzer.analyze(ExpressionAnalyzer.java:258)
	at com.facebook.presto.sql.analyzer.ExpressionAnalyzer.analyzeExpression(ExpressionAnalyzer.java:1466)
	at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.analyzeExpression(StatementAnalyzer.java:1834)
	at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.analyzeSelect(StatementAnalyzer.java:1646)
	at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.legacyVisitQuerySpecification(StatementAnalyzer.java:909)
	at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitQuerySpecification(StatementAnalyzer.java:854)
	at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitQuerySpecification(StatementAnalyzer.java:242)
	at com.facebook.presto.sql.tree.QuerySpecification.accept(QuerySpecification.java:127)
	at com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)
	at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:254)
	at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:264)
	at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitQuery(StatementAnalyzer.java:605)
	at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitQuery(StatementAnalyzer.java:242)
	at com.facebook.presto.sql.tree.Query.accept(Query.java:94)
	at com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)
	at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:254)
	at com.facebook.presto.sql.analyzer.StatementAnalyzer.analyze(StatementAnalyzer.java:228)
	at com.facebook.presto.sql.analyzer.Analyzer.analyze(Analyzer.java:72)
	at com.facebook.presto.sql.analyzer.Analyzer.analyze(Analyzer.java:64)
	at com.facebook.presto.execution.SqlQueryExecution.doAnalyzeQuery(SqlQueryExecution.java:305)
	at com.facebook.presto.execution.SqlQueryExecution.analyzeQuery(SqlQueryExecution.java:291)
	at com.facebook.presto.execution.SqlQueryExecution.start(SqlQueryExecution.java:247)
	at com.facebook.presto.execution.QueuedExecution.lambda$start$1(QueuedExecution.java:62)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)
```

`row('a')::row(varchar(1))` should implicitly cast to `row(varchar(2))`

**Comment by electrum** [comment]
Some test cases for `TestTypeRegistry`:
```java
assertCommonSuperType("row(a bigint,b bigint)", "row(a bigint,b bigint)", "row(a bigint,b bigint)");
assertCommonSuperType("row(a int,b bigint)", "row(a bigint,b bigint)", "row(a bigint,b bigint)");
assertCommonSuperType("row(a int,b bigint)", "row(a bigint,b int)", "row(a bigint,b bigint)");

**Comment by shixuan-fan** [comment]
Any idea about rows with the same type but different name? For example, `row(a bigint)` and `row(b bigint)`.

**Comment by shixuan-fan** [comment]
Addressed by #9035
